chainer/training/updater.py

```diff
@@ -0,0 +1,366 @@
+
+        """
+        raise NotImplementedError
+
+    def update(self):
+        """Updates the parameters of the target model.
+
+        This method implements an update formula for the training task,
+        including data loading, forward/backward computations, and actual
+        updates of parameters.
+
+        This method is called once at each iteration of the training loop.
+
+        """
+        raise NotImplementedError
+
+    def serialize(self, serializer):
+        """Serializes the current state of the updater object."""
+        raise NotImplementedError
+
+
+class StandardUpdater(Updater):
+
+    """Standard implementation of Updater.
+
+    This is the standard implementation of :class:`Updater`. It accepts one or
+    more training datasets and one or more optimizers. The default update
+    routine assumes that there is only one training dataset and one optimizer,
+    while users can specify their own update routines. Each batch is converted
+    to input arrays by :func:`~chainer.datasets.concat_examples` by default,
+    which can also be manually set.
+
+    There are two ways to modify the update behavior besides setting a custom
+    loss function. One is by setting a custom update function via the
+    ``update_func`` argument. The other one is by inheriting this class and
+    overriding the :meth:`update` method. In latter case, do not forget to
+    update the iteration counter at each call of this method, because this
+    value is watched by the trainer for deciding when to invoke extensions and
+    when to exit the training loop.
+
+    Args:
+        iterator: Dataset iterator for the training dataset. It can also be a
+            dictionary of iterators. If this is just an iterator, then the
+            iterator is registered by the name ``'main'``.
+        optimizer: Optimizer to update parameters. It can also be a dictionary
+            of optimizers. If this is just an optimizer, then the optimizer is
+            registered by the name ``'main'``.
+        update_func: Update routine. This is a function that takes the updater
+            object as the argument. The default routine uses ``converter`` and
+            ``loss_func`` if specified.
+        converter: Converter function to build input arrays. Each batch
+            extracted by the main iterator and the ``device`` option are passed
+            to this function. If it is omitted,
+            :func:`~chainer.dataset.concat_examples` is used. If
+            ``update_func`` is specified, this argument is ignored and not
+            used.
+        device: Device to which the training data is sent. Negative value
+            indicates the host memory (CPU). If ``update_func`` is specified,
+            this argument is ignored and not used.
+        loss_func: Loss function. The target link of the main optimizer is used
+            by default. If ``update_func`` is specified, this argument is
+            ignored and not used.
```

Comment:
Hum, it's clearer. In my opinion, a closure is simple but a little confusing. A class is complex but easier to read. 

BTW, why do you make a `_update_func` an attribute? I feel simply users need to override `update` method of `Updater`. Is it not enough?